### ISSUE TITLE
Custom DNS wildcard for subdomains only

### ISSUE THREAD

**Issue by starsoccer** [0xERR0R/blocky#1515]
From reading the documentation it seems like its currently possible to specify a domain and have all subdomains for it go to a single IP for example purposes Ill use 192.168.178.3 like the documentation. The documentation uses an example of `printer.lan` and then `printer.lan`, `my.printer.lan` or `i.love.my.printer.lan` will all resolve to the same IP(192.168.178.3).

I cant seem to find any documentation on how to specify an IP for all subdomains except the top level. For instance Id like to be able to do something like this in my config file:
`
*.printer.lan: 192.168.178.3
`

The above seems to be invalid, but I would expect it to do something like this:
`
printer.lan -> go to upstream dns
my.printer.lan -> 192.168.178.3
something.printer.lan -> 192.168.178.3
i.love.my.printer.lan -> 192.168.178.3
`

Basically the goal being to have the top level domain still go to the upstream resolver but any subdomains to go to the hard coded IP.

**Comment by bcookatpcsd** [comment]
I literally think that is the exact example.. 

https://0xerr0r.github.io/blocky/latest/configuration/#custom-dns

```
   customDNS:
       customTTL: 300s
       filterUnmappedTypes: true
       mapping:
         pool.ntp.org: 172.16.254.241
```

 drill -Q i.love.my.pool.ntp.org @10.120.0.247
172.16.254.241

(I do not see an * in the documentation.. you shouldn't have one either.. )

**Comment by starsoccer** [comment]
@bcookatpcsd So in your example I want `pool.ntp.org` to continue to go to its regular IP however its normally resolved, but I want `my.pool.ntp.org`, as well as `i.love.my.pool.ntp.org` to go to 172.16.254.241.

Basically I want the top level domain to continue to be resolved like normal, but I want all subdomains to be resolved differently.

**Comment by bcookatpcsd** [comment]
I would start with this.. and see what you get then get to work.. 

https://0xerr0r.github.io/blocky/latest/configuration/#conditional-dns-resolution

or just hard code the records you want from the other domain and serve them as your own..

**Comment by starsoccer** [comment]
> I would start with this.. and see what you get then get to work..
> 
> https://0xerr0r.github.io/blocky/latest/configuration/#conditional-dns-resolution
> 
> or just hard code the records you want from the other domain and serve them as your own..

Im not sure I am following exactly on how I would use the conditional DNS to achieve what I want?

Regarding hard coding the records, that is probably the easiest option and is what I am currently doing. I just wanted to avoid the extra step of having to add every subdomain to blocky for it to be resolved. I was hoping there would be a way to just resolve all subdomains to one IP and then the top level fall back to the main resolver since the IP can change.

**Comment by starsoccer** [comment]
Bumping

**Comment by polarathene** [comment]
## Resolve "wildcard" locally (_resolve top-level via upstreams_)

> Basically I want the top level domain to continue to be resolved like normal, but I want all subdomains to be resolved differently.

Here is how you can query everything but the top-level domain itself to a resolve a specific IP. For convenience it's a Docker Compose `compose.yaml` with the blocky config embedded.

```yaml
services:
  dns:
    image: ghcr.io/0xerr0r/blocky
    container_name: dns-blocky
    ports:
      - "53:53/tcp"
      - "53:53/udp"
    configs:
      - source: blocky-config
        target: /app/config.yml

configs:
  blocky-config:
    content: |
      # Any query not handled by `customDNS` is queried through the upstreams:
      upstreams:
        groups:
          default:
            - 1.1.1.1 # Cloudflare

      # `example.com` => query upstream
      # `*.example.com` => 192.168.0.42
      customDNS:
        # Avoid querying upstreams if record type is missing:
        filterUnmappedTypes: true
        # All queries to `example.com` (except top-level) are rewritten to `example.internal`:
        # hello.world.example.com => hello.world.example.internal => 192.168.0.42
        rewrite:
          example.com: example.internal
        # All queries to `example.internal` (including top-level) will respond with this IP:
        mapping:
          example.internal: 192.168.0.42

# Required when the host bind port is already in use such as systemd-resolved at `127.0.0.53:53`.
# This is an alternative to changing the default host binding from `0.0.0.0` to `127.0.0.1`,
# instead of publishing each port with the explicit IP prefix (like `127.0.0.1:53:53/tcp`)
networks:
  default:
    name: example-net
    driver_opts:
      com.docker.network.bridge.host_binding_ipv4: 127.0.0.1
```

---

## Conditional DNS (_route query to another DNS service_)

> I'm not sure I am following exactly on how I would use the conditional DNS to achieve what I want?

Conditional DNS that was suggested isn't necessary if you're using `customDNS` like above for Blocky to respond to such queries.

If you instead required resolving via a different DNS server that wasn't blocky or an upstream, you would use `conditional` config similar to `customDNS` logic above but instead of an IP to return it would forward the query to that alternative DNS service to respond back with.

```yaml
# Add this to /app/config.yml

conditional:
  # Avoid sending query to upstreams when it was conditionally matched here but failed:
  fallbackUpstream: false
  # Same logic as described with `customDNS.rewrite` + `customDNS.mapping`:
  rewrite:
    example.com: example.internal
    localhost: example-net
  # However instead of returning the IP assigned, the query is forwarded to 127.0.0.1 (blocky in this case),
  # which will receive `hello.world.example.internal`.
  # This can be useful with Docker embedded DNS at 127.0.0.11:53 for it's private bridge networks.
  mapping:
    example.internal: 127.0.0.1 # Blocky DNS (query again with a potentially rewritten query)
    example-net: 127.0.0.11 # Embedded Docker DNS (dns-blocky.localhost => dns-blocky.example-net)
    .: 127.0.0.11 # Likewise forward any single-label queries to Docker DNS (eg: dns-blocky)
```

In this example, to make it a bit more interesting I added an example using the earlier config `networks.default.name: example-net` + `services.dns.container_name: dns-blocky`, so that a query could come from outside of that docker private network (_such as via the host_), but be routed to Docker embedded DNS to respond with the container IP associated to that private network.

**Comment by samascience** [comment]
this would make it block ads.instagram.com but allow instagram.com !! hopefully which the pihole can't achieve !!

**Comment by garfieldairlines** [comment]
Bot for the bump

**Comment by NextBlaubeere** [comment]
Commenting, because i ran into the same issue today.

Wildcards would be neat to put into mappings.

**Comment by unlair** [comment]
Thanks @polarathene! I tried out your suggestion and it worked perfectly. Maybe this could be put in the docs somewhere as the current workaround as it's probably a common use-case.
